src/test/scala/dsptools/numbers/FixedPointSpec.scala

```diff
@@ -26,26 +28,109 @@ class FixedRing1(val width: Int, val binaryPoint: Int) extends Module {
+
+  def expectedValue(value: Double, left: Boolean, shift: Int): Double = {
+    val factor = 1 << c.binaryPoint
+    val x = value * factor
+    val y = x.toInt
+    val z = if(left) y << shift else y >> shift
+    val w = z.toDouble / factor
+    w
+  }
+
+  def truncate(value: Double): Double = {
+    val coeff = 1 << c.binaryPoint
+    val x = value * coeff
+    val y = x.toInt.toDouble
+    val z = y / coeff
+    z
+  }
+  poke(c.io.dynamicShiftValue, 0)
+
+  private val (minSIntValue, maxSIntValue) = firrtl_interpreter.extremaOfSIntOfWidth(c.width)
+
+  private val minValue = minSIntValue.toDouble * increment
+  private val maxValue = maxSIntValue.toDouble * increment
+
+  for(value <- minValue to maxValue by increment) {
+    poke(c.io.inValue, value)
+    expect(c.io.shiftLeftResult, expectedValue(value, left = true, c.fixedShiftSize),
+      s"shift left ${c.fixedShiftSize} of $value should be ${expectedValue(value, left = true, c.fixedShiftSize)}")
+    expect(c.io.shiftRightResult, expectedValue(value, left = false, c.fixedShiftSize),
+      s"shift right ${c.fixedShiftSize} of $value should be ${expectedValue(value, left = false, c.fixedShiftSize)}")
+
+    step(1)
+
+    for(dynamicShiftValue <- 0 until c.width) {
+      poke(c.io.dynamicShiftValue, dynamicShiftValue)
+      step(1)
+      expect(c.io.dynamicShiftLeftResult, expectedValue(value, left = true, dynamicShiftValue),
+        s"dynamic shift left $dynamicShiftValue of $value should " +
+          "be ${expectedValue(value, left = true, dynamicShiftValue)}")
+      expect(c.io.dynamicShiftRightResult, expectedValue(value, left = false, dynamicShiftValue),
+        s"dynamic shift right $dynamicShiftValue of $value should" +
+          s"be ${expectedValue(value, left = false, dynamicShiftValue)}")
+    }
+  }
+}
 
 class FixedPointSpec extends FreeSpec with Matchers {
-  "FixedPointIsReal functions should work for a number of different precisions" - {
-    for(binaryPoint <- 0 to 4) {
-      s"should work with binary point $binaryPoint" in {
-        dsptools.Driver.execute(() => new FixedRing1(16, binaryPoint = binaryPoint)) { c =>
-          new FixedRing1Tester(c)
-        } should be (true)
+  "FixedPoint numbers should work properly for the following mathematical type functions" - {
+    "The ring family" - {
+      for (binaryPoint <- 0 to 4) {
+        s"should work, with binaryPoint $binaryPoint" in {
+          dsptools.Driver.execute(() => new FixedRing1(16, binaryPoint = binaryPoint)) { c =>
+            new FixedRing1Tester(c)
+          } should be(true)
+        }
+      }
+    }
+
+    "The shift family" - {
+      for (binaryPoint <- 0 to 16 by 2) {
+        s"should work with binary point $binaryPoint, for positives even when binaryPoint > width" in {
+          dsptools.Driver.execute(
+            () => new FixedPointShifter(8, binaryPoint = binaryPoint, fixedShiftSize = 2),
+            Array("-tdb", "2")
```

Comment:
Also double check with Verilator -- that gets things down the full pipe. FirrtlInterpreter wouldn't catch any translation issues from low Firrtl to Verilog, right?